### DISCUSSION TITLE
Need help with a custom logger function to send logs through a slack webhook

### DISCUSSION THREAD

**Discussion by nickolasdeluca** [fastify/fastify#4944] [Q&A] [answered]
I wrote a custom logger based on [this](https://stackoverflow.com/questions/55264854/how-can-i-use-custom-logger-in-fastify) answer and here's the code: 

```javascript
import https from "https";

class SlackLogger {
  constructor(...args) {
    this.args = args;
  }

  info(message) {
    sendSlackMessage(
      JSON.stringify({
        text: message,
      })
    );
  }

  error(message) {
    sendSlackMessage(
      JSON.stringify({
        text: message,
      })
    );
  }

  debug(message) {
    sendSlackMessage(
      JSON.stringify({
        text: message,
      })
    );
  }

  fatal(message) {
    sendSlackMessage(
      JSON.stringify({
        text: message,
      })
    );
  }

  warn(message) {
    sendSlackMessage(
      JSON.stringify({
        text: message,
      })
    );
  }

  trace(message) {
    sendSlackMessage(
      JSON.stringify({
        text: message,
      })
    );
  }

  child(message) {
    return new SlackLogger(message);
  }
}

function sendSlackMessage(message) {
  let url =
    "myurl";

  const options = {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.parse(message),
  };

  const req = https.request(url, options, (res) => {
    res.on("data", (d) => {});
  });

  req.on("error", (error) => {});

  req.write(message);
  req.end();
}

export default SlackLogger;
```

This is how I attach my custom function to my fastify server:
```javascript
import SlackLogger from "./logger/slack.js";
const logger = new SlackLogger();

import Fastify from "fastify";
const fastify = Fastify({
  logger: env === "production" ? logger : true,
});
```
It works flawlessly with the first logging info, i.e: 
```shell
Server listening at http://[::1]:8001

Server listening at http://127.0.0.1:8001/
```
But when I make a get/post request to my server, it the code throws this error:
```shell
JSON.stringify({
       ^
TypeError: Converting circular structure to JSON
```
I understand why, the code is trying to parse into a string a structure which is circular, it's understandable.
What I can't find is how to get the final message like those that are shown in the console when you run fastify with the default logger, i.e: 
```shell
{"level":30,"time":1690845595856,"pid":202208,"hostname":"myhost","msg":"Server listening at http://[::1]:8001"}
{"level":30,"time":1690845595857,"pid":202208,"hostname":"myhost","msg":"Server listening at http://127.0.0.1:8001"}
{"level":30,"time":1690845597970,"pid":202208,"hostname":"myhost","reqId":"req-1","req":{"method":"GET","url":"/public/ping","hostname":"localhost:8001","remoteAddress":"::1","remotePort":64437},"msg":"incoming request"}
{"level":30,"time":1690845597974,"pid":202208,"hostname":"myhost","reqId":"req-1","res":{"statusCode":200},"responseTime":3.333199977874756,"msg":"request completed"}
```

Anyone able to show me how to get the final parsed string which would be perfect for me to send to my slack channel? I don't want any other information, what is provided is more than enough to find and debug errors and stuff like that.

Any help will be much appreciated.

**Comment by metcoder95** [comment]
Hey 👋 
Just to double-check if I understood correctly. Does the problem is how to print messages likes the one shown, not the circular reference issue?

**Reply by nickolasdeluca** [reply]
> Hey 👋 
> 
> Just to double-check if I understood correctly. Does the problem is how to print messages likes the one shown, not the circular reference issue?

Yes.

**Comment by Eomm** [comment, marked as answer]
I was thinking that a transport could be better and I found this one: https://www.npmjs.com/package/pino-slack-transport

By using pino, you will get log at full speed

**Reply by nickolasdeluca** [reply]
Thanks for the suggestion.
Just implemented it and works flawlessly.
It's a shame there's little room for customization. I'll try to implement a custom solution based on this package later, but it'll serve me well until then.

(the doc could be better tho, jesus, had to hack both examples together to make it work lmao)
